Add parent loggers to autocompletion

daniel-beck opened the pull request:
> ![screen shot](https://user-images.githubusercontent.com/1831569/37295830-78f32270-2619-11e8-8a55-21e009c60693.png)

> ![screen shot](https://user-images.githubusercontent.com/1831569/37295913-b8a34c4c-2619-11e8-9a08-3e72b7f3461f.png)


Subsumes https://github.com/jenkinsci/jenkins/pull/3344 while at the same time doing a LOT more work 😆 

### Proposed changelog entries

Too minor.


### Submitter checklist

- [n/a] JIRA issue is well described
- [n/a] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [n/a] Appropriate autotests or explanation to why this change has no tests

daniel-beck committed:
Add parent loggers to autocompletion

Wadeck commented on core/src/main/java/hudson/logging/LogRecorder.java:
```diff
@@ -87,14 +88,21 @@
     @Restricted(NoExternalUse.class)
     public AutoCompletionCandidates doAutoCompleteLoggerName(@QueryParameter String value) {
-        AutoCompletionCandidates candidates = new AutoCompletionCandidates();
+        List<String> candidateNames = new ArrayList<>();
```
nit: could be a `Set` in order to optimize the `candidateNames.contains(...)` but as the list will not have lots of element (except for single letter search or "org", "com", etc.) the user will not see any difference.

dwnusbaum replied:
If it was a set you could remove the `!candidateNames.contains(loggerNamePrefix)` check entirely, but I agree that it shouldn't really make a difference.

Wadeck reviewed (approved):
:bee: LGTM, I close mine.

Wadeck commented:
After some tests, it seems that the autocomplete does not provide consistency among time / user interaction. I mean that if I type `org.random` slowly, as `org` will return lots of results compared to `org.random` (empty normally), potentially the empty result will arrive at the client before the `org` and when it will receive the `org` result, it will display it, even if it's not the last thing the user typed.

daniel-beck commented:
I really don't know whether this is a net improvement…

![screen shot](https://user-images.githubusercontent.com/1831569/37351262-f2b1ebc0-26da-11e8-9777-95a7e182b465.png)

daniel-beck committed:
Use a LinkedHashSet to maintain insertion order while saving contains check

Wadeck commented:
> I really don't know whether this is a net improvement…

In terms of UX, I think the user is more confident if he wants to use "org" and see it in the list explicitely. So it's a +1 from me.

Concerning the order, I think the insertion order is not relevant since the `LogManager.getLogManager().getLoggerNames()` is not sorted.

![screenshot_2018-03-13_163117_001](https://user-images.githubusercontent.com/2662497/37351739-0ace5b20-26dc-11e8-866b-431033762e02.png)

daniel-beck commented:
@Wadeck Yes, I've learned that now as well. I'll investigate this when I'm next bored on the weekend :-)

daniel-beck committed:
More improvements to logging autocomplete

daniel-beck committed:
Suggest all parent loggers that are relevant

Parent loggers will be suggested in auto completion iff they
include two or more child, grand-child, etc. loggers, and there
is no more specific parent logger that includes the same number
of descendants.

daniel-beck commented:
I spend a bit of time making this actually nice yesterday.

Current state:

> ![screen shot](https://user-images.githubusercontent.com/1831569/37521222-321e4cbc-2920-11e8-8b1b-1326439ad4b1.png)

No loggers in `org.jenkinsci.main` that aren't in `org.jenkinsci.main.modules`, so don't suggest that.

daniel-beck commented:
Major rework since previous reviews.

Wadeck commented on core/src/main/java/hudson/logging/LogRecorder.java:
```diff
@@ -85,16 +87,60 @@
 
+    @Restricted(NoExternalUse.class)
+    @VisibleForTesting
+    public static Set<String> getAutoCompletionCandidates(List<String> loggerNamesList) {
```
As the order is kept, I think it could be nice to either explicit that in the name of the method or return a `SortedSet`

Wadeck commented on core/src/main/java/hudson/logging/LogRecorder.java:
```diff
@@ -85,16 +87,60 @@
+        // Example: 'org' will show 'org', because there's org.apache, org.jenkinsci, etc.
+        // 'io' might only show 'io.jenkins.plugins' rather than 'io' if all loggers starting with 'io' start with 'io.jenkins.plugins'.
+        HashMap<String, Integer> seenPrefixes = new HashMap<>();
+        SortedSet<String> relevantPrefixes = new TreeSet<>();
```
Normally I would have suggested to use `new TreeSet<>(String.CASE_INSENSITIVE_ORDER)`, but the separation between the pure package name and the Logger name is somewhat nice. 👍 

I hope we will never have package with uppercased name or class with lowercased name.

oleg-nenashev replied:
> I hope we will never have package with uppercased name or class with lowercased name. 

IIRC we have some such packages in the code, but I do not care much about such case. Performance is better with the standard ordering

Wadeck commented on core/src/main/java/hudson/logging/LogRecorder.java:
```diff
@@ -85,16 +87,60 @@
+        }
+
+        // get names of all actual loggers known to Jenkins
+        Set<String> candidateNames = new LinkedHashSet<>(getAutoCompletionCandidates(Collections.list(LogManager.getLogManager().getLoggerNames())));
```
perhaps split into two lines for readability ?

Wadeck reviewed (approved):
Only small comment, nothing blocking.

> No loggers in org.jenkinsci.main that aren't in org.jenkinsci.main.modules, so don't suggest that.

Nice UX improvement, good job 👍 

+ thank you for the test, that avoid me to run the code with special case to be it works as expected :)

oleg-nenashev commented on core/src/main/java/hudson/logging/LogRecorder.java:
```diff
@@ -85,16 +87,60 @@
+
+            String longerPrefix = null;
+            for (int i = loggerNameParts.length; i > 0; i--) {
+                String loggerNamePrefix = StringUtils.join(Arrays.copyOf(loggerNameParts, i), ".");
```
Sound like not the best implementation for high-depth classes
